samplers/samplers.go

```diff
@@ -72,6 +73,131 @@ type JSONMetric struct {
 	Value []byte `json:"value"`
 }
 
+// CardinalityCountName is the name of the metric that we eventually pass on
+// to the sinks.
+const CardinalityCountName = "cardinality.count"
+
+// CardinalityCountType is the string that denotes the type of the metric. We
+// use this when we serialize to a JSON metric.
+const CardinalityCountType = "cardinality_count"
+
+// cardinalityPrecision is the precision of the HLL that stores cardinality for
+// each metric name.
+const cardinalityPrecision = 18
+
+// CardinalityCount is a sampler that records an approximate cardinality count
+// by metric name.
+type CardinalityCount struct {
+	MetricCardinality map[string]*hyperloglog.HyperLogLogPlus
+}
+
+func NewCardinalityCount() *CardinalityCount {
+	return &CardinalityCount{make(map[string]*hyperloglog.HyperLogLogPlus)}
+}
+
+// Sample adds a measurement to the cardinality counter. It makes the
+// assumption that no two metrics with different types share a name (ie, name
+// is unique across all types)
+func (cc *CardinalityCount) Sample(name string, joinedTags string) {
+	if _, present := cc.MetricCardinality[name]; !present {
+		hll, _ := hyperloglog.NewPlus(cardinalityPrecision)
+		cc.MetricCardinality[name] = hll
+	}
+	hasher := fnv.New64a()
+	hasher.Write([]byte(joinedTags))
+	cc.MetricCardinality[name].Add(hasher)
+}
+
+// defin a custom error type here
+type CardinalityExportError struct {
+	metricNames   []string
+	errorMessages []string
+}
+
+func (e *CardinalityExportError) Error() string {
+	return "oh no"
+}
+
+// an array of names
+// an arr of the actual error message
+
+// ExportSets breaks apart the CardinalityCount into individual JSON metrics
+// by metric name.
+func (cc *CardinalityCount) ExportSets() ([]JSONMetric, error) {
+	jsonMetrics := make([]JSONMetric, 0, len(cc.MetricCardinality))
+
+	for metricName, hll := range cc.MetricCardinality {
+
+		buf := &bytes.Buffer{}
+		encoder := gob.NewEncoder(buf)
```

Comment:
We recently had #220 show up which got us into a bit of trouble because the transmission of this sort of data isn't versioned or anything. I get a bit worried when I see new gob-encoded things. This is probably not actionable _now_ but something we should consider for the future. Should we version the import endpoint or something?